### DISCUSSION TITLE
Supporting rfft~ and rifft~

### DISCUSSION THREAD

**Discussion by jameysharp** [Wasted-Audio/hvcc#99] [General]
I heard from https://hackers.town/@JoshuaACNewman that there are open questions about [how to implement FFT support in hvcc](https://hackers.town/@JoshuaACNewman/110012001165891611). I don't know anything about PD or Heavy, but I do know a thing or two about FFTs, dataflow languages, and compilers, and wanted to offer some help. I don't want to hang out in a Discord though so I thought I'd see if folks would be interested in having a conversation here.

I'm having trouble making sense of Heavy's `ir2c` generator, but if I'm reading it correctly, it looks like the main difficulty is that every signal block takes one sample at a time, or at most, one SIMD vector of samples at a time. That would be troublesome since an FFT necessarily has to wait for N samples to arrive before it can produce any output, where N is the length of the FFT.

I see from the example of `SignalConvolution` that a signal block can keep a ring-buffer of past inputs in local state. I think that pattern should work for FFTs as well: every time an input sample arrives, save it in the ring buffer, and when the ring buffer wraps, compute the FFT or inverse FFT of the whole buffer.

The other difficulty is it looks like signal blocks are synchronous: they must produce an output sample for every input sample. A length-N FFT introduces N samples worth of latency, so there's no meaningful value to output for the first N samples. That's okay as long as PD's semantics allow emitting N samples of zeroes at the beginning. Does it?

If there aren't any other challenges I've missed, then I think this is… relatively straightforward, and I'd be happy to chat further about how to make it work.

**Comment by dromer** [comment]
Thank you a lot for the input @jameysharp !
Do note that Heavy does not contain a line of PD code. Essentially it emulates a PD DSP graph using its own internal library.

Very cool to hear that this has some possibility to be implemented, however even though I took over maintenance of the project I know very little about the internal workings of Heavy, or about programming DSP code in general. So any help with this is certainly appreciated.

**Reply by dromer** [reply]
Btw I was looking at https://community.vcvrack.com/t/complete-list-of-c-c-fft-libraries/9153 for potential libraries that could help with this. Main requirements would be: written in C and compatible license to Heavy. And of course performance and ease of integration would help.

**Reply by jameysharp** [reply]
The only FFT library I've personally used is FFTW, which I gather isn't a good choice for Heavy's embedded targets. Joshua had mentioned you were looking at pffft, so I glanced at it briefly. It seems probably fine?

I mentioned wondering what PD does because if you're trying to emulate its behavior, it's important to know what that behavior is. Maybe you can answer some questions for me about that.

First, is there something like a "length" input or config option to the `rfft~` signal? It should be a positive integer. For best performance ideally it's a power of 2.

Can that length change while the graph is running? I'm hoping it's a static property of the graph, and doesn't change without stopping and rebuilding the dataflow graph. I'm not entirely sure how to make a synchronous dataflow graph deal with a dynamically changing FFT length.

When one floating-point sample arrives at input to `rfft~`, does it produce one floating-point sample of output? Or does PD allow a signal block to produce a length-N array of samples all at once after receiving N inputs? I'm guessing it's one sample in, one sample out.

When you start up a patch with an FFT in it, what are the first values that come out? I'm guessing you get N zeros at the beginning, where N is the length parameter.

I'll write down some advice here assuming that I've guessed correctly on the above questions.

I'd start by copying the ir2c files with `SignalConvolution` in their names to `SignalFFT` or something. There might be better examples but that's the one I looked at and I think it should serve well enough.

During initialization, allocate one buffer that can hold N samples, where N is the FFT length parameter. (I think Heavy calls these buffers "tables" for some reason?) I'm assuming that every element of the buffer will be initialized to zero but you might need to do that yourself. Also store an index into the buffer, initialized to 0. Finally, you probably need to initialize your FFT library as well so it can precompute some constants; if you use pffft, call `pffft_new_setup(N, PFFFT_REAL)` and save the pointer it returns alongside the buffer and index.

When you receive a sample of input (or a SIMD vector of samples), find the buffer element at the current index. Copy it out to a temporary variable: This will be your output sample. Then copy your input into that element of the buffer. Increment the index. If the index is now equal to the length of the buffer, it's time to do an FFT and then reset the index to zero.

When that happens, call whatever FFT implementation you've chosen. If that's pffft, I think you want `pffft_transform_ordered`. That library's documentation says that "input and output may alias", which is convenient: it means that you can pass a pointer to your buffer for both the `input` and `output` parameters, and the result will be placed right back into the same buffer. That's what we want and it's why I said above that you should output samples from the same buffer that you're writing the input into.

`pffft_transform_ordered` also needs a temporary buffer passed in for the `work` parameter. The documentation says that you can pass this as `NULL` for FFTs of size less than 16,384, but that advice means it'll allocate up to 64kB on the stack, which I suspect is a bad plan on your embedded targets. I think you might actually be better off using `pffft_aligned_malloc` and `pffft_aligned_free` for every call, or you could also try allocating work space once at initialization time. You may have to experiment with all three options and see what the trade-offs are between memory usage and time spent in the memory allocator.

Note that all of the above works exactly the same for both forward and backward/inverse FFTs. The main difference is that when you call `pffft_transform_ordered`, you'll pass either `PFFFT_FORWARD` (for `rfft~`) or `PFFFT_BACKWARD` (for `rifft~`) for the `direction` parameter. The other thing is that for `rifft~` you need to divide each output sample by N because many FFT libraries, pffft included, don't normalize the result.

I don't have time to write the implementation of this myself, but I can certainly try to answer questions. Also, if you want, you can put up a PR and tag me and I'll be happy to review it. I might be able to find some potential problems even though I don't know much about Heavy.

**Reply by JoshuaACNewman** [reply]
This is awesome. Thank you for helping the project like this.

**Reply by dromer** [reply]
@jameysharp I've actually never used these objects myself. For posterity here the internal help file from Pure Data.
That should give you at least some idea of how these are typically used:

![](https://i.imgur.com/lQ4rbEZ.png)

So it seems there is no length parameter. Instead we always get one block.

Note that `block~` and `switch~` are not used in Heavy. Its runtime target determines the fixed blocksize.

By default I think the size is set to 256 samples: https://github.com/Wasted-Audio/hvcc/blob/develop/hvcc/generators/ir2c/HeavyTable.py#L59

**Reply by jameysharp** [reply]
That diagram is very helpful, thank you!

It looks like the block size is necessary for the FFT signals because there seems to be no other way to set the size of the FFT. Also I think that means they are _not_ padding the output with zeroes at the beginning, but relying on everything being buffered to the block size.

Can you require Heavy's users to provide extra information about the FFT length to use? I think there are good reasons to use longer block sizes anyway but it would be nice to get something minimally working without needing to rework block sizes everywhere.

Interesting to see that they did not normalize the inverse FFT, so you shouldn't do the thing I said about dividing by N.

**Reply by dromer** [reply]
Even though the context process() function receives the number of frames (blocksize), right now I do not see a way to retrieve this information so we can use that value.

@jameysharp what do you mean with "provide extra information"? The way the blocksize is set by the runtime is through the heavy context interface: https://github.com/Wasted-Audio/hvcc/blob/develop/hvcc/generators/ir2c/static/HeavyContextInterface.hpp#L115

One issue we may need to think about is that a DAW (vst2/3/lv2/clap for instance) may change the blocksize while the dsp is running, this means the associated table also needs to be resized (and probably causes other issues?). I don't know if that is possible.

**Reply by dromer** [reply]
I did a little test by adding a
```cpp
sendFloatToReceiver(0xB5B01859, static_cast<float>(n4)); // send to __hv_blocksize
```
to the main `process()` function and this will send the current actual block size to a special receiver object that we can query to at least set normalization. Possibly we can query this receiver from the rfft~ functions as well.

**Reply by dromer** [reply]
@jameysharp I'm staring a bit more at this now (and putting some first tries in generating code for it) but I'm a bit confused to how we can get separate real and imaginary parts out of (and into) `pffft_transform_ordered`.

**Reply by dromer** [reply]
Ah, from these notes it appears that the imaginary part is interleaved and there are two possible layouts for this (for which you suggested the ordered one): https://chromium.googlesource.com/chromium/src/+/master/third_party/pffft/README.md

So I suppose that will need some manual (un)tangling to set up the buffer correctly (I have no experience with this).

**Reply by dromer** [reply]
@jameysharp when you have time please take a look at https://github.com/Wasted-Audio/hvcc/pull/110 which should hopefully have most parts needed. Still very much WIP (and pseudo) on the HvSignalRFFT code.

**Reply by dromer** [reply]
Reading back to what you said before @jameysharp "I'm guessing it's one sample in, one sample out." I believe this depends on the SIMD implementation. The buffers that are used can be either 1, 4 or 8 value structures.

I indeed noticed that `pffft_transform_ordered()` requires single float values, so I assume that we have to iterate over this array and run it through the fft function?

**Comment by dromer** [comment]
I took a bit of a dive into the internals and `SignalConvolution` is actually related to an implementation of a `conv~` object which does not exist in Pure Data, but (together with `lorenz~`) has some heavy-specific code. It's unknown to me where it originates from.

It will take some piecing together from various other objects to better understand how to implement `rfft~/rifft~` further "up" the chain.

**Comment by dromer** [comment]
Potential implementation from Emilie Gillet that at least works on embedded/STM32 target (MIT): <https://github.com/electro-smith/DaisyExamples/blob/master/field/Nimbus/shy_fft.h>

**Reply by falkTX** [reply]
that doesnt look suitable for dpf/plugins, as it relies on input vs output buffers being different.
on quite a few hosts they are the same, reused for saving on cpu, often called "inplace processing".

I have used pffft before, can recommend it. supports both neon and sse, plus has a license compatible with DPF and commercial plugins

**Reply by dromer** [reply]
Do you think pffft is also suitable for STM32?

**Reply by falkTX** [reply]
that I have no idea, was mostly thinking about audio plugins here
since pffft is just common C code, I dont see why not

**Comment by dromer** [comment]
Some more docs on rfft~ usage in pd: http://www.pd-tutorial.com/english/ch03s08.html
